### PR TITLE
Reading collection from server

### PR DESCRIPTION
**Pull request by denarch22**
## 📝 Associated issues
[Traer colecciones desde el servidor stac](https://app.zenhub.com/workspaces/all-5f7f96a92b54ca00239a9348/issues/gh/pem-humboldt/biotablero-search-backend/26)

### CODE REVIEW

File: app/services/read_coleccion.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+from typing import Optional, Dict, Any
```
Comment (ManuelStardust) [inline app/services/read_coleccion.py]: I think it's better to just call this file "collection"

File: app/routes/metrics.py [hunk 2]
```diff
@@ -81,15 +81,15 @@ async def get_areas_by_defined_area(
 
 @router.post("/{metric_id}/areas", response_model=List[AreasResponse])
 async def get_areas_by_polygon(
-    metric_id: Annotated[str, fastapi.Depends(metric_id_param)],
-    polygon: Polygon,
-) -> list[dict[str, float]]:
+    metric_id: str, polygon: Polygon
```
Comment (ManuelStardust) [inline app/routes/metrics.py]: It is better to keep the type that metric_id previously had. Or do you think there is some reason why it needs to be only string?

File: app/routes/metrics.py [hunk 3]
```diff
@@ -81,15 +81,15 @@ async def get_areas_by_defined_area(
...
+) -> dict[str, Any]:
```
Comment (ManuelStardust) [inline app/routes/metrics.py]: Is there any reason to put an Any type instead of float or number?

**Review by ManuelStardust — CHANGES_REQUESTED** [review]

**Comment by denarch22** [comment]
the adjustments indicated in the comments are made

**Comment by denarch22** [comment]
I had to make this adjustment: return {"metric_id": metric_id}
) -> str:
return metric_id , since when passing the variable in the find_collection_url method, it is expecting a string and not a dictionary, with this it works in the return of the list, I make the pr request again for review.

File: requirements.txt [hunk 4]
```diff
@@ -6,4 +6,4 @@ geojson-pydantic>=1.0.1,<1.0.2
 rio-tiler>=6.6.1,<6.7.0
 rasterio>=1.3.9,<1.3.10
 geopandas>=1.0.0,<1.0.1
-shapely>=2.0.3,<2.0.4
+shapely>=2.0.3,<2.0.4
```
Comment (erikasv) [inline requirements.txt]: please leave the new line at the end of the file

File: requirements.txt
Comment (erikasv) [inline requirements.txt:1]: when running the app I'm getting `ModuleNotFoundError: No module named 'requests'`

File: app/services/collection.py
Comment (erikasv) [inline app/services/collection.py:1]: I think this file should be inside `services/utils`

File: app/services/metrics.py [hunk 5]
```diff
@@ -19,30 +17,60 @@ def get_areas_by_defined_area(area_type, area_id):
...
+        collection_url = find_collection_url(settings.stac_url, metric_id)
+        if not collection_url:
+            raise ValueError(f"No collection found for id: {metric_id}")
+
+        items_url = get_collection_items_url(collection_url)
+
+        if not items_url:
+            raise ValueError(
+                f"No items URL found for collection id: {metric_id}"
+            )
```
Comment (erikasv) [inline app/services/metrics.py]: these lines are repeated in both functions, and will be repeated in the services for defined areas too, the logic to get the items for a given collection should be handled in the `collection` file.

File: app/utils/config.py [hunk 6]
```diff
@@ -7,6 +7,7 @@ class Settings(BaseSettings):
 
     stac_url: str = "http://localhost:8080"
     env: str = "dev"
+    cog_base_url: str = "https://staccatalog.blob.core.windows.net/cog-test"
```
Comment (erikasv) [inline app/utils/config.py]: this variable is not used

File: app/services/collection.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    root_data = response.json()
+
+    for link in root_data.get("links", []):
+        href = link.get("href")
+        if (
+            isinstance(href, str)
+            and link.get("rel") == "child"
+            and collection_id in href
+        ):
+            return href
```
Comment (erikasv) [inline app/services/collection.py]: although this logic is correct, I don't see a reason to no request directly `{base_url}/collections/{collection_id}` if the collection doesn't exist it will return an error

File: app/services/collection.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+def get_collection_items_url(collection_url: str) -> Optional[str]:
+    response = requests.get(collection_url)
+    response.raise_for_status()
+    collection_data = response.json()
+
+    for link in collection_data.get("links", []):
+        if link.get("rel") == "items":
+            return link.get("href")
+
```
Comment (erikasv) [inline app/services/collection.py]: The same as the previous function but with `/items` at the end

**Review by erikasv — CHANGES_REQUESTED** [review]

**Comment by denarch22** [comment]
The respective adjustment of the functions is carried out according to the comments made.

File: app/routes/metrics.py [hunk 9]
```diff
@@ -45,11 +45,11 @@ class AreasResponse(BaseModel):
 
 async def metric_id_param(
     metric_id: Annotated[
-        Literal["LossPersistence", "Coverage"],
-        fastapi.Path(description="metric you whish to query"),
+        str,
```
Comment (erikasv) [inline app/routes/metrics.py]: why did you change it?

Comment (denarch22) [reply app/routes/metrics.py]: What happens is that when performing tests to see if the collection is found or not, I identify that what this does is burn the name of the metric_id, meaning that if I place another metric in the endpoint, it will always indicate that it is not there because it is passing that burned parameter.

Comment (erikasv) [reply app/routes/metrics.py]: that's te point. The backend will not process requests for metrics that don't exist

File: app/services/metrics.py [hunk 10]
```diff
@@ -19,30 +17,35 @@ def get_areas_by_defined_area(area_type, area_id):
...
 
-    def get_areas_by_polygon(polygon):
-        # TODO: read raster from STAC
-        raster_cloud_path = "https://staccatalog.blob.core.windows.net/cog-test/Colombia_pp-2015_12_31-pp_2011_2015.tif"
+    def get_areas_by_polygon(
+        polygon: PolygonGeometry, metric_id: str
+    ) -> list[dict[str, float]]:
+        collection_url = find_collection_url(settings.stac_url, metric_id)
+
+        items_url = get_collection_items_url(collection_url)
```
Comment (erikasv) [inline app/services/metrics.py]: I still don't see why it's necessary to have the collection url first and then the items, why don't just get the items directly?

File: app/services/utils/collection.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+            and collection_id in href
+        ):
+            collection_url = href
+            break
+
+    if not collection_url:
+        raise ValueError(
+            f"No collection found for id at URL: {base_url}/collections/{collection_id}"
+        )
+
```
Comment (erikasv) [inline app/services/utils/collection.py]: why did you just dismiss the previous comment? I don't see a reason to keep looking in the links object instead of just querying the direct collection url

Comment (erikasv) [reply app/services/utils/collection.py]: the same for the items

File: requirements.txt [hunk 12]
```diff
@@ -7,3 +7,4 @@ rio-tiler>=6.6.1,<6.7.0
 rasterio>=1.3.9,<1.3.10
 geopandas>=1.0.0,<1.0.1
 shapely>=2.0.3,<2.0.4
+requests>=2.31.0,<2.32.0
```
Comment (erikasv) [inline requirements.txt]: keep the new line

**Review by erikasv — CHANGES_REQUESTED** [review]

**Comment by denarch22** [comment]
The changes were made as previously discussed, I remain attentive.

File: app/services/metrics.py [hunk 13]
```diff
@@ -19,30 +13,29 @@ def get_areas_by_defined_area(area_type, area_id):
...
 
-    def get_areas_by_polygon(polygon):
-        # TODO: read raster from STAC
-        raster_cloud_path = "https://staccatalog.blob.core.windows.net/cog-test/Colombia_pp-2015_12_31-pp_2011_2015.tif"
+    def get_areas_by_polygon(
+        polygon: PolygonGeometry, metric_id: str
+    ) -> list[dict[str, float]]:
+
+        first_asset = get_first_item_asset_from_collection(
+            settings.stac_url, metric_id
```
Comment (erikasv) [inline app/services/metrics.py]: I think it's better to read stac_url setting directly from collection

File: app/services/metrics.py [hunk 14]
```diff
@@ -19,30 +13,29 @@ def get_areas_by_defined_area(area_type, area_id):
...
+        )
+        raster_cloud_path = first_asset
```
Comment (erikasv) [inline app/services/metrics.py]: This reassignment is not necessary

File: app/services/utils/collection.py [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+import requests
+
+
+def get_collection_items_url(
+    base_url: str, collection_id: str
+) -> Optional[str]:
+    url = base_url + collection_id
+    response = requests.get(url)
+    response.raise_for_status()
```
Comment (erikasv) [inline app/services/utils/collection.py]: Instead of just raising the error verify if it's 404 and return a user friendly message (the same as you had before)

File: app/services/utils/collection.py [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    collection_data = response.json()
+
+    items_url = None
+    for link in collection_data.get("links", []):
+        if link.get("rel") == "items":
+            items_url = link["href"]
+            break
```
Comment (erikasv) [inline app/services/utils/collection.py]: again, don't search through the links, but instead verify directly the items url

File: app/routes/metrics.py (hunk 9 above)
Comment (erikasv) [reply app/routes/metrics.py]: you haven't addressed this point

**Review by erikasv — CHANGES_REQUESTED** [review]

**Comment by denarch22** [comment]
Adjustments are made based on the latest comments made

File: app/services/utils/collection.py (hunk 15 above)
Comment (erikasv) [reply app/services/utils/collection.py]: this function will raise 404 errors but won't have the message, look if it's possible to adjust the message, or just handle it manually

File: app/services/utils/collection.py [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+def get_collection_items_url(collection_id: str) -> Optional[str]:
+    url = settings.stac_url + collection_id
+    response = requests.get(url)
+    response.raise_for_status()
+    collection_data = response.json()
+
+    if not collection_data:
+        raise ValueError(f"No collection found at URL: {url}")
+
+    return url
```
Comment (erikasv) [inline app/services/utils/collection.py]: at this moment the function is returning the collection url not the items url. You can either change the function name or check for the items in this same function and return the items url

File: app/services/metrics.py [hunk 18]
```diff
@@ -19,30 +13,23 @@ def get_areas_by_defined_area(area_type, area_id):
...
-
-        collection_metric = next(
-            (
-                collection
-                for collection in collections
-                if collection["id"] == metric_id
-            ),
-            None,
-        )
+    def get_layer_by_polygon(metric_id: str, polygon: PolygonGeometry):
```
Comment (erikasv) [inline app/services/metrics.py]: I just noticed these two functions have the parameters in different order (get_areas_by_polygon and get_layer_by_polygon, please leave metric_id first in both

**Review by erikasv — CHANGES_REQUESTED** [review]

**Comment by denarch22** [comment]
Adjustments are made based on the latest comments made

File: app/services/metrics.py [hunk 19]
```diff
@@ -1,16 +1,10 @@
...
-        "id": "LossPersistence",
-        "cog": "https://staccatalog.blob.core.windows.net/cog-test/Colombia_pp-2015_12_31-pp_2011_2015.tif",
-    },
-    {
-        "id": "Coverage",
-        "cog": "https://coverage-example.tif",
-    },
-]
+
+settings = config.get_settings()
```
Comment (erikasv) [inline app/services/metrics.py]: settings isn't being used

File: app/services/utils/collection.py [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+                detail=f"HTTP error occurred while accessing {items_url}",
+            )
+
+    items_data = response.json()
+
+    if not items_data.get("features"):
+        raise HTTPException(
+            status_code=404,
+            detail=f"No items found in the collection at URL: {items_url}",
+        )
```
Comment (erikasv) [inline app/services/utils/collection.py]: delete these, the same validation is done in load_first_item_asset

**Review by erikasv — CHANGES_REQUESTED** [review]

**Review by erikasv — APPROVED** [review]

**Review by ManuelStardust — APPROVED** [review]
